lib: allow cpp program to call strlcpy, strlcat and explicit_bzero

nanfengnan1 opened the pull request.

Jafaral reviewed (changes requested):
I do see that you have an open issue related to this change. Can you please link it in the commit message, and also add a proper description? 

also, change the prefix "lib:" to "build:" in the commit message since that is a change to the build system, it is not a change to the lib code.

Commit:
build: allow external cpp program to call strlcpy, strlcat and explicit_bzero

issue link: https://github.com/FRRouting/frr/issues/20199

It enables external C++ programmes to correctly parse the C symbol table without utilising C++'s name qualification mechanism.

Signed-off-by: alexan <fenglei@genew.com>

eqvinox commented on configure.ac:
```diff
@@ -1405,22 +1405,46 @@ dnl yes this might include stddef.h 3 times but who cares
 #include <stddef.h>
-size_t strlcpy(char *__restrict dest,
-	       const char *__restrict src, size_t destsize);
+size_t strlcpy(char *dest, const char *src, size_t size);
```
C++ has no `restrict`

eqvinox replied:
I don't think we should fix C++ at the cost of C here though, removing the `restrict` lessens the optimizations the compiler can do.

eqvinox commented on configure.ac:
```diff
@@ -1405,22 +1405,46 @@ dnl yes this might include stddef.h 3 times but who cares
 AH_VERBATIM(HAVE_STRLCPY_, [
 #ifndef HAVE_STRLCPY
+#ifdef __cplusplus
+extern "C" {
```
`extern "C"` should never wrap `#include` statements (e.g. `stddef.h` below)

eqvinox reviewed (changes requested):
I'm gonna throw up a somewhat better fix for this in a minute.

nanfengnan1 commented:
I think we should close it if exist better pr

eqvinox commented:
PR superseded ages ago
